### PR TITLE
feat(one-time backup): part-2 backup app metadata and send rpc to partition primary

### PR DESCRIPTION
**Pull request by zhangyifan27**
Core steps for a one-time cold backup:
- meta server handle rpc and init backup. #772 
- **backup app metadata and send rpc to partition primary.**
- handle backup response and complete a backup.

### CODE REVIEW

**Review by hycdong — APPROVED** [review]

File: src/meta/backup_engine.cpp [hunk 1]
```diff
@@ -74,9 +75,128 @@ error_code backup_engine::set_block_service(const std::string &provider)
...
+    backup_policy_info.__set_policy_name(get_policy_name());
+    req->policy = backup_policy_info;
+    req->backup_id = _cur_backup.backup_id;
+    req->app_name = _cur_backup.app_name;
+
+    ddebug_f("backup_id({}): send backup request to partition {}, target_addr = {}",
+             _cur_backup.backup_id,
+             pid.to_string(),
+             partition_primary.to_string());
+    backup_rpc rpc(std::move(req), RPC_COLD_BACKUP, 0_ms, 0, pid.thread_hash());
```
Comment (neverchanje) [inline src/meta/backup_engine.cpp:164]: You still need to set a timeout for this RPC and do retry-sort-of-thing if the RPC is unresponsive. Consider giving it 10s.

Comment (zhangyifan27) [reply src/meta/backup_engine.cpp]: Already set a 10s timeout, `on_backup_reply` would handle rpc errors.

**Review by hycdong — APPROVED** [review]

File: src/replica/backup/replica_backup_server.h [hunk 2]
```diff
@@ -20,13 +20,13 @@
 #include <dsn/dist/replication/replication_types.h>
 #include <dsn/cpp/rpc_holder.h>
 
+#include "common/replication_common.h"
```
Comment (levy5307) [inline src/replica/backup/replica_backup_server.h:23]: Why did you move backup_rpc to `replication_common.h`

File: src/meta/backup_engine.h [hunk 3]
```diff
@@ -60,18 +60,29 @@ class backup_engine
...
-    error_code get_app_stat(int32_t app_id, std::shared_ptr<app_state> &app);
+    error_code write_backup_file(const std::string &file_name, const dsn::blob &write_buffer);
+    error_code backup_app_meta();
+    void backup_app_partition(const gpid &pid);
+    void on_backup_reply(error_code err,
+                         const backup_response &response,
+                         gpid pid,
+                         const rpc_address &primary);
+
+    std::string get_policy_name() const
```
Comment (levy5307) [inline src/meta/backup_engine.h:74]: `const std::string get_policy_name() const`

File: src/meta/backup_engine.h [hunk 4]
```diff
@@ -60,18 +60,29 @@ class backup_engine
...
+    }
 
     backup_service *_backup_service;
     dist::block_service::block_filesystem *_block_service;
     std::string _provider_type;
     dsn::task_tracker _tracker;
 
     // lock the following variables.
-    dsn::zlock _lock;
+    mutable dsn::zlock _lock;
```
Comment (levy5307) [inline src/meta/backup_engine.h:85]: Why did you make _lock mutable?

File: src/meta/backup_engine.cpp [hunk 5]
```diff
@@ -74,9 +71,140 @@ error_code backup_engine::set_block_service(const std::string &provider)
...
+    {
+        zauto_read_lock l;
+        _backup_service->get_state()->lock_read(l);
+        std::shared_ptr<app_state> app = _backup_service->get_state()->get_app(_cur_backup.app_id);
+        if (app == nullptr || app->status != app_status::AS_AVAILABLE) {
+            derror_f("app {} is not available, couldn't do backup now.", _cur_backup.app_id);
+            return ERR_INVALID_STATE;
+        }
+        app_state tmp = *app;
+        // do not write app envs to backup file
```
Comment (levy5307) [inline src/meta/backup_engine.cpp:118]: I think It's better to add a explanation to show why don't write app envs to backup file

File: src/replica/backup/replica_backup_server.h (hunk 2 above)
Comment (zhangyifan27) [reply src/replica/backup/replica_backup_server.h:23]: Seems that most rpc_hoder defined in `replication_common.h` or `replica_stub.h`.

File: src/meta/backup_engine.h (hunk 4 above)
Comment (zhangyifan27) [reply src/meta/backup_engine.h:85]: To make `is_backing_up()` const.

**Review by neverchanje — APPROVED** [review]

File: src/meta/backup_engine.cpp [hunk 6]
```diff
@@ -74,9 +71,140 @@ error_code backup_engine::set_block_service(const std::string &provider)
...
+        zauto_read_lock l;
+        _backup_service->get_state()->lock_read(l);
+        std::shared_ptr<app_state> app = _backup_service->get_state()->get_app(_cur_backup.app_id);
+        if (app == nullptr || app->status != app_status::AS_AVAILABLE) {
+            derror_f("app {} is not available, couldn't do backup now.", _cur_backup.app_id);
+            return ERR_INVALID_STATE;
+        }
+        app_state tmp = *app;
+        // Because we don't restore app envs, so no need to write app envs to backup file.
+        tmp.envs.clear();
```
Comment (hycdong) [inline src/meta/backup_engine.cpp:119]: I think it may be better to contain app envs in files, otherwise, app envs will be lost if the old table is dropped. App envs will include many useful data, such as throttle, org, access control etc. If you are sure this information is useless, you can remove it.

**Review by hycdong — APPROVED** [review]
